File: srclab-common/src/main/java/xyz/srclab/common/util/proxy/ProxyClass.java
package xyz.srclab.common.util.proxy;

import xyz.srclab.annotation.Immutable;

@Immutable
public interface ProxyClass<T> {

    static <T> ProxyClassBuilder<T> newBuilder(Class<T> type) {
        return ProxyClassSupport.newBuilder(type);
    }

    T newInstance();

    T newInstance(Class<?>[] parameterTypes, Object[] args);

    Class<T> getProxyClass();
}

File: boat-common/src/main/kotlin/xyz/srclab/common/proxy/ProxyClass.kt
package xyz.srclab.common.proxy

import xyz.srclab.common.base.Current

/**
 * @author sunqian
 */
interface ProxyClass<T : Any> {

    fun newInstance(): T

    fun newInstance(parameterTypes: Array<Class<*>>?, args: Array<Any?>?): T

    //fun getProxyClass(): Class<T>

    companion object {

        @JvmStatic
        fun <T : Any> newProxyClass(proxyClass: Class<T>, proxyMethods: Iterable<ProxyMethod<T>>): ProxyClass<T> {
            return newProxyClass(Current.classLoader, proxyClass, proxyMethods)
        }

        @JvmStatic
        //@JvmOverloads
        fun <T : Any> newProxyClass(
            classLoader: ClassLoader = Current.classLoader,
            proxyClass: Class<T>,
            proxyMethods: Iterable<ProxyMethod<T>>
        ): ProxyClass<T> {
            return ProxyClassGenerator.DEFAULT.generate(classLoader, proxyClass, proxyMethods)
        }
    }
}